rpc/grpc/server/services/nodeservice/service.go

```diff
@@ -0,0 +1,201 @@
+package nodeservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	nodesvc "github.com/cometbft/cometbft/api/cometbft/services/node/v1"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/p2p"
+	"github.com/cometbft/cometbft/rpc/core"
+)
+
+// server implements nodesvc.NodeServiceServer.
+// The node service is a gRPC endpoint serving requests for information about
+// the CometBFT node provinding the gRPC interface.
+type server struct {
+	log     log.Logger
+	nodeEnv *core.Environment
+}
+
+// New returns a gRPC server serving requests for information about a CometBFT node.
+func New(l log.Logger, env *core.Environment) nodesvc.NodeServiceServer {
+	return &server{
+		log:     l.With("service", "NodeService"),
+		nodeEnv: env,
+	}
+}
+
+// GetStatus is the gRPC endpoint serving requests for the node current status.
+// This includes the node's basic info, such as public key, latest block hash, app
+// hash, block height, and time.
+// The request object isn't used in the current implementation, and the function
+// doesn't expect clients to provide any data in it.
+func (s *server) GetStatus(
+	ctx context.Context,
+	_ *nodesvc.GetStatusRequest,
+) (*nodesvc.GetStatusResponse, error) {
+	l := s.log.With("endpoint", "GetStatus")
+
+	if ctx.Err() != nil {
+		err := ctx.Err()
+
+		l.Error("exited early because of context cancellation", "err", err)
+
+		formatStr := "client canceled request: %s"
+		return nil, status.Errorf(codes.Canceled, formatStr, err)
+	}
+
+	nodeInfo, err := s.collectNodeInfo()
+	if err != nil {
+		l.Error(err.Error(), "err", err)
+
+		clientErrMSg := "node's basic information unavailable"
+		return nil, status.Error(codes.Internal, clientErrMSg)
+	}
```

Comment:
Is there a standard we follow to return errors from APIs?
For example, here (and below as well) I log the detailed error, but return to the client a more generic error that masks what really happened internally. In this particular example, the error the function logs is something like `p2p.NodeInfo concrete type != p2p.DefaultNodeInfo: <weird_type_name_here>`

The assumption is that clients don't need to know or wouldn't understand the details of the error. Is it useful information for a client to know that the node's code is using an unexpected concrete type for the interface `p2p.NodeInfo`?

In short, I'm wondering if we make this sort of distinction on error returns, or if we simply return the "unfiltered" error to clients.

Reply:
We should log an error message that the operator can send to implementers for troubleshooting. When it comes to clients, we should provide a custom error message that gives an idea of what caused the error, but doesn't reveal too many internal details.